experimental/devices/mfrc522/mfrc522.go

```diff
@@ -0,0 +1,752 @@
+// Copyright 2018 The Periph Authors. All rights reserved.
+// Use of this source code is governed under the Apache License, Version 2.0
+// that can be found in the LICENSE file.
+
+// Package mfrc522 controls a Mifare RFID card reader.
+//
+// The code is largely ported from Python library : https://github.com/mxgxw/MFRC522-python
+//
+// The datasheet is available at : https://www.nxp.com/docs/en/data-sheet/MFRC522.pdf
+package mfrc522
+
+import (
+	"errors"
+	"fmt"
+	"periph.io/x/periph/conn/gpio"
+	"periph.io/x/periph/conn/gpio/gpioreg"
+	"periph.io/x/periph/conn/spi"
+	"periph.io/x/periph/experimental/devices/mfrc522/commands"
+	"time"
+)
+
+// Dev is an handle to an MFRC522 RFID reader.
+type Dev struct {
+	resetPin      gpio.PinOut
+	irqPin        gpio.PinIn
+	Authenticated bool
+	maxSpeedHz    int64
+	spiDev        spi.Conn
+}
+
+// DefaultKey provides the default bytes for card authentication for method B.
+var DefaultKey = []byte{0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF}
+
+const maxSpeed = int64(1000000)
+
+/*
+NewSPI creates and initializes the RFID card reader attached to SPI.
+	spiPort - the SPI device to use.
+	resetPin - reset GPIO pin.
+	irqPin - irq GPIO pin.
+*/
+func NewSPI(spiPort spi.Port, resetPin, irqPin string) (*Dev, error) {
+
+	spiDev, err := spiPort.Connect(maxSpeed, spi.Mode0, 8)
+	if err != nil {
+		return nil, err
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	dev := &Dev{
+		spiDev:     spiDev,
+		maxSpeedHz: maxSpeed,
+	}
+
+	pin := gpioreg.ByName(resetPin)
+	dev.resetPin = pin
+	dev.resetPin.Out(gpio.High)
+
+	pin = gpioreg.ByName(irqPin)
+	dev.irqPin = pin
+	dev.irqPin.In(gpio.PullUp, gpio.FallingEdge)
+
+	err = dev.Init()
+
+	return dev, nil
+}
+
+var initCommands = [][]byte{
+	{commands.TModeReg, 0x8D},
+	{commands.TPrescalerReg, 0x3E},
+	{commands.TReloadRegL, 30},
+	{commands.TReloadRegH, 0},
+	{commands.TxAutoReg, 0x40},
+	{commands.ModeReg, 0x3D},
+}
+
+// Init initializes the RFID chip.
+func (r *Dev) Init() error {
+	err := r.Reset()
```

Comment:
if err := r.Reset(); err != nil {

Do the same at lines 87 and 92, which will save 3 lines and make the function looks much more trivial.

Since this device is effectively a memory mapped register, you can coalesce the commands TModeReg, TPrescalerReg, TReloadRegH, TReloadRegL as a single write. The problem is that you call them "commands" but they are effectively memory addresses.

So basically, initCommands would look like:

```
var initCommands = [][]byte{
  // Mode, Prescaler, ReloadH, ReloadL
  {0x2A, 0x8B, 0x3E, 0x00, 0x30},
  {commands.TxAutoReg, 0x40},
  {commands.ModeReg, 0x3D},
}
```

Makes sense?

More reference at https://periph.io/x/periph/conn/mmr

Reply:
Why do you think that it is memory-mapped?

Reply:
See pages 10~11 of the datasheet.

Reply:
Pages 36~37 present the Register addresses and each register is described up to page 69. In this parlance, register ~= memory. This is not "commands" per se, but writing to a register may induce an operation, which is alike sending a command. But strictly speaking, you are simply writing a value in a registry at the specified address.

Formal commands are described at page 70~73 and they are executed by writing to the pseudo-register "CommandReg". This is the only real "command" that there is.

Reply:
I think that it's better to postpone this update for the next iteration, because I'll need some more time to figure out how to deal with the memory mapping registers in this case. I'll focus on making the code more "standard", and then we can get back to this topic once I finish with the rest. The device is functional and works just fine with the current way of the communication.

will that work?

Reply:
Yes this is fine.